### PR TITLE
`Development`: Move translation keys for error messages into error.json

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
I noticed that there are two different locations in the current codebase where translations for error messages were stored. One was `error.json`, the other `global.json`. 

### Description
We should only have one general location for error messages. Therefore I moved all strings from `global.json` to `error.json`.

This PR additionally fixes a small bug with programming exercises and auxiliary repositories. The description was not sent to the server and therefore did not get saved.

### Steps for Testing
Try to invoke some server errors and check that the resulting alert shows correct strings in both German and English.

Additionally, test the saving of descriptions for auxiliary repositories:

Create and / or edit a programming exercise:

1. Add a new auxiliary repository to the exercise and set values for all fields.
2. Save the exercise and check that all values got saved.
3. Change the description to a long text (> 500 chars)
4. Make sure that a translated error message is shown. This message should also contain the maximum length of 500 chars.

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by Dominik-Weinzierl — APPROVED** [review]
Code change look good to me, but we need to verify that those errors work still as expected - i am not sure if there was a reason for this decision :thinking:

File: src/main/webapp/i18n/de/error.json [hunk 1]
```diff
@@ -47,6 +47,32 @@
         "submissionPolicyToggleFailedPolicyNotExist": "Für diese Programmieraufgabe konnte die Abgaberichtlinie nicht (de)aktiviert werden, da die Programmieraufgabe keine Abgaberichtlinie hat. ",
         "submissionPolicyActiveNull": "Abgaberichtlinien müssen aktiviert oder deaktiviert sein. Dieser Status darf nicht null sein.",
         "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!",
-        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest."
+        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest.",
+        "internalServerError": "Interner Serverfehler",
+        "server.not.reachable": "Server ist nicht erreichbar",
+        "url.not.found": "Nicht gefunden",
+        "NotNull": "Feld {{fieldName}} kann nicht leer sein!",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:54]:
```suggestion
        "notNull": "Feld {{fieldName}} darf nicht leer sein!",
```
Watch out, I also changed "NotNull" to "notNull"

File: src/main/webapp/i18n/de/error.json [hunk 2]
```diff
@@ -47,6 +47,32 @@
...
+        "Size": "Feld {{fieldName}} erfüllt nicht die minimalen/maximalen Voraussetzungen!",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:55]:
```suggestion
        "size": "Feld {{fieldName}} hat nicht die minimalen/maximalen Längenanforderungen!",
```
Watch out, I also changed "Size" to "size"

Maybe also include the minimum and maximum length?

File: src/main/webapp/i18n/de/error.json [hunk 3]
```diff
@@ -47,6 +47,32 @@
...
+        "userexists": "Login bereits vergeben!",
+        "emailexists": "E-Mail wird bereits verwendet!",
+        "idexists": "Ein neuer {{entityName}} kann noch keine ID haben",
+        "ambiguousContext": "Ein neuer {{entityName}} muss einen eindeutigen Kontext haben",
+        "invalid.template.build.plan.id": "Die Vorlagen-Build-Plan-ID scheint ungültig zu sein. Stelle bitte sicher, dass sie existiert und für Artemis zugänglich ist.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:60]: If you don't change the keys to camelCase, you could also extract the invalid into a subcategory for all following keys with "invalid."

**Review by JohannesStoehr — COMMENTED** [review]
The code changes look good. 
But I'd suggest improving the translations and the structure of the json as well. I left some comments about my suggestions:

File: src/main/webapp/i18n/de/error.json [hunk 4]
```diff
@@ -47,6 +47,32 @@
         "submissionPolicyToggleFailedPolicyNotExist": "Für diese Programmieraufgabe konnte die Abgaberichtlinie nicht (de)aktiviert werden, da die Programmieraufgabe keine Abgaberichtlinie hat. ",
         "submissionPolicyActiveNull": "Abgaberichtlinien müssen aktiviert oder deaktiviert sein. Dieser Status darf nicht null sein.",
         "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!",
-        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest."
+        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest.",
+        "internalServerError": "Interner Serverfehler",
+        "server.not.reachable": "Server ist nicht erreichbar",
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/webapp/i18n/de/error.json:52]: These keys are generated by Angular and have to stay like this.

File: src/main/webapp/i18n/de/error.json (hunk 2 above)
Comment (Strohgelaender) [reply src/main/webapp/i18n/de/error.json:55]: I also think that these keys are part of some internal error messages. They were generated in the initial setup in 2016.

File: src/main/webapp/i18n/de/error.json [hunk 5]
```diff
@@ -47,6 +47,40 @@
...
+        "url.not.found": "Nicht gefunden",
+        "NotNull": "Feld {{fieldName}} darf nicht leer sein!",
+        "size": "Feld {{fieldName}} erfüllt nicht die minimalen/maximalen Längenanforderungen!",
+        "userexists": "Login bereits vergeben!",
+        "emailexists": "E-Mail wird bereits verwendet!",
+        "idexists": "Ein neuer {{entityName}} kann noch keine ID haben",
+        "ambiguousContext": "Ein neuer {{entityName}} muss einen eindeutigen Kontext haben",
+        "missingname": "Ein neues {{entityName}} braucht einen Namen",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:60]: I did not find any use of that key inside of Artemis. Is this also uses by jhipster or something similar?

File: src/main/webapp/i18n/de/error.json [hunk 6]
```diff
@@ -47,6 +47,40 @@
...
-        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest."
+        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest.",
+        "internalServerError": "Interner Serverfehler",
+        "server.not.reachable": "Server ist nicht erreichbar",
+        "url.not.found": "Nicht gefunden",
+        "NotNull": "Feld {{fieldName}} darf nicht leer sein!",
+        "size": "Feld {{fieldName}} erfüllt nicht die minimalen/maximalen Längenanforderungen!",
+        "userexists": "Login bereits vergeben!",
+        "emailexists": "E-Mail wird bereits verwendet!",
+        "idexists": "Ein neuer {{entityName}} kann noch keine ID haben",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:58]: These keys don't use camelCase and are also used in the server as keys.
Most other keys are in camelCase

File: src/main/webapp/i18n/de/error.json [hunk 7]
```diff
@@ -47,6 +47,40 @@
...
+            "auxiliary.repository": {
+                "id": "Die ID eines Hilfs-Repositories scheint gesetzt zu sein. Stelle bitte sicher, dass die ID nicht gesetzt ist.",
+                "name": "Der Name von mindestens einem Hilfs-Repository ist ungültig oder dupliziert. Stelle bitte sicher, dass jeder Name gültig ist und es keine Duplikate gibt.",
+                "checkout.directory": "Das Checkout-Verzeichnis von mindestens einem Hilfs-Repository ist ungültig oder dupliziert. Stelle bitte sicher, dass jedes Checkout-Verzeichnis gültig ist und es keine Duplikate gibt.",
+                "description": "Die Beschreibung von mindestens einem Hilfs-Repository länger als maximal {{maxLength}} erlaubte Zeichen!"
+            },
+            "testcases.weights": "Vor Deaktivierung der manuellen Bewertung muss für mindestens einen Testfall die Bewertungsgewichtung größer Null gesetzt werden.",
+            "page.request": "Die Seitenanforderung scheint ungültig zu sein. Stelle bitte sicher, dass genügend {{entityName}}-Entitäten vorhanden sind, damit die Anforderungsseite erstellt werden kann."
+        },
+        "idnull": "Ungültige ID",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/error.json:79]: Also all lower case and used as key in java classes

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Some more nitpicks

File: src/main/webapp/i18n/de/error.json (hunk 5 above)
Comment (Dominik-Weinzierl) [reply src/main/webapp/i18n/de/error.json:60]: Some error messages are implicit and handled automatically - there is no explicit call on the client needed (e.g. see delete user)

Comment (Dominik-Weinzierl) [reply src/main/webapp/i18n/de/error.json:60]: image

:thinking:

File: src/main/webapp/i18n/de/error.json [hunk 8]
```diff
@@ -47,6 +47,39 @@
...
+            },
+            "solution": {
+                "build.plan.id": "Die Lösungs-Build-Plan-ID scheint ungültig zu sein. Stelle bitte sicher, dass sie existiert und für Artemis zugänglich ist.",
+                "repository.url": "Die Lösungs-Repository-URL scheint ungültig zu sein. Stelle bitte sicher, dass sie existiert und für Artemis zugänglich ist."
+            },
+            "auxiliary.repository": {
+                "id": "Die ID eines Hilfs-Repositories scheint gesetzt zu sein. Stelle bitte sicher, dass die ID nicht gesetzt ist.",
+                "name": "Der Name von mindestens einem Hilfs-Repository ist ungültig oder dupliziert. Stelle bitte sicher, dass jeder Name gültig ist und es keine Duplikate gibt.",
+                "checkout.directory": "Das Checkout-Verzeichnis von mindestens einem Hilfs-Repository ist ungültig oder dupliziert. Stelle bitte sicher, dass jedes Checkout-Verzeichnis gültig ist und es keine Duplikate gibt.",
+                "description": "Die Beschreibung von mindestens einem Hilfs-Repository länger als maximal {{maxLength}} erlaubte Zeichen!"
```
Comment (Dominik-Weinzierl) [inline src/main/webapp/i18n/de/error.json:73]: This sentence sounds weird :D

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html [hunk 9]
```diff
@@ -90,7 +90,7 @@ <h4 *ngIf="isImport" id="jhi-programming-exercise-heading-import" jhiTranslate="
                 </ngx-datatable-column>
                 <ngx-datatable-column name="Description" prop="description">
                     <ng-template ngx-datatable-cell-template let-row="row">
-                        <textarea class="form-control table-editable-field__input me-2" [id]="row.id + '-description'" [(ngModel)]="row.description"> </textarea>
+                        <textarea class="form-control table-editable-field__input me-2" [id]="row.id + '-description'" [(ngModel)]="row.description" name="description"> </textarea>
```
Comment (Dominik-Weinzierl) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html:93]: One question: Why is this change needed? :thinking: 

Edit: Seems like thats the bug fix right?

**Review by Dominik-Weinzierl — COMMENTED** [review]
Code looks good to me - two suggestions and questions :+1:

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html (hunk 9 above)
Comment (Strohgelaender) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html:93]: Yes, that was needed in order to save the description correctly.

**Review by rhubinger — DISMISSED** [review]
Tested on TS2. Works as expected

**Review by Marvin-Vitcu — DISMISSED** [review]
Code changes look good.

**Review by MetinLamby — APPROVED** [review]
Error messages "The description of at least one Auxiliary Repository is too long!" and "Die Beschreibung von mindestens einem Hilfs-Repository ist zu lang!" appeared as expected

**Comment by SimonPaul02** [comment]
image
These words don't get translated. (But not really part of the PR)

**Review by SimonPaul02 — DISMISSED** [review]
Tested on Ts2, works as expected

**Review by yuling-sun — DISMISSED** [review]
Tested on TS2. Works as described.

**Review by nairaarnez — APPROVED** [review]
Tested in TS1, the message displayed, the message is also correct (with "ist")

**Review by JohannesStoehr — DISMISSED** [review]
Code changes look good after the merge

**Review by ge65cer — APPROVED** [review]
Generally I'm against stacking unrelated changes in the same PR, but since your fix is simply adding a `name` attribute to the HTML tag I'd make an exception.

Extracting the translations from `global.json` is a welcome refactor. LGTM.

**Review by JohannesStoehr — DISMISSED** [review]
Rechecked the lastest merge commit. Code still looks good

**Review by jpbernius — APPROVED** [review]

**Comment by jpbernius** [comment]
LGTM. @Strohgelaender can you resolve the merge conflict? I would merge this today.

**Review by manuelmanso — APPROVED** [review]
LGTM

**Review by ge65cer — APPROVED** [review]
Re-approve

**Comment by ge65cer** [comment]
Ping @jpbernius

**Review by ge65cer — APPROVED** [review]
